safehttp/plugins/xsrf/xsrf.go

```diff
@@ -0,0 +1,172 @@
+	// TODO(@mihalimara22): make endpoints a variadic of string literals
+	ad := map[string]bool{}
+	for _, p := range domains {
+		ad[p] = true
+	}
+	return &Plugin{
+		appKey:       appKey,
+		s:            s,
+		fmPolicyMode: FMEnforceMode,
+		allowedCORS:  ad,
+	}
+}
+
+// SetFetchMetadataPolicyMode allows setting the Fetch Metadata policy, enforced
+// by default to one of the other options. Setting it to report will allow
+// requests that violate the policy to pass but will log violation. Setting it
+// to disable
+func (p *Plugin) SetFetchMetadataPolicyMode(mode string) error {
+	switch mode {
+	case FMEnforceMode, FMDisabledMode, FMReportMode:
+		p.fmPolicyMode = mode
+	default:
+		return errors.New("invalid Fetch Metadata policy")
+	}
+	return nil
+}
+
+// GenerateToken generates a cryptographically safe XSRF token using the user ID
+// and the path.
+func (p *Plugin) GenerateToken(host string) (string, error) {
+	userID, err := p.s.GetUserID()
+	if err != nil {
+		return "", fmt.Errorf("token generation failed: %v", err)
+	}
+	token := xsrftoken.Generate(p.appKey, userID, host)
+	return token, nil
+}
+
+// validateToken validates the XSRF token. This should be present in all
+// requests as the value of form parameter xsrf-token.
+func (p *Plugin) validateToken(ir *safehttp.IncomingRequest) (safehttp.StatusCode, bool) {
+	userID, err := p.s.GetUserID()
+	if err != nil {
+		return safehttp.Status401Unauthorized, false
+	}
+	// TODO(@mihalimara22): add multipart support as well
+	f, err := ir.PostForm()
+	if err != nil {
+		return safehttp.Status400BadRequest, false
+	}
+	token := f.String(TokenKey, "")
+	if f.Err() != nil || token == "" {
+		return safehttp.Status403Forbidden, false
+	}
+	ok := xsrftoken.Valid(token, p.appKey, userID, ir.Header.Get("Host"))
+	if !ok {
+		return safehttp.Status403Forbidden, false
+	}
+	return 0, true
+}
+
+// applyPolicy validates the request using the default Fetch Metadata policy and
+// only allows it to pass if it conforms to the policy. A cross-origin request
+// will not be allowed unless targeted to a domain that is allowed to server
+// cross-origin. The policy will not be applied if the browser doesn't have
+// Fetch Metadata support implemented.
+// TODO(@empijei, @kele, @mattiasgrenfeldt, @mihalimara22): decide whether the user should be allowed to provide
+// their own policy
+func (p *Plugin) applyPolicy(ir *safehttp.IncomingRequest) (safehttp.StatusCode, bool) {
+	switch ir.Header.Get("Sec-Fetch-Site") {
+	case "":
+		// Fetch Metadata not supported by the browser
+		return 0, true
+	case "same-origin", "same-site", "none":
+		return 0, true
+	case "cross-origin":
+		if p.allowedCORS[ir.Header.Get("Host")] {
+			return 0, true
+		}
+		return safehttp.Status403Forbidden, false
+	}
+
+	switch ir.Header.Get("Sec-Fetch-Mode") {
+	case "navigate":
+		dest := ir.Header.Get("Sec-Fetch-Dest")
+		if dest == "object" || dest == "embed" {
```

Comment:
These two lines can be combined:
```go
if dest := ...; dest == ... || dest == ... { 
```
There are also some `if`s in `Before` where this can be applied.